numpyro/infer/mcmc.py

```diff
@@ -615,33 +687,74 @@ def __init__(self,
         if (chain_method == 'parallel' and num_chains > 1) or (
                 "CI" in os.environ or "PYTEST_XDIST_WORKER" in os.environ):
             self.progress_bar = False
-
+        self._jit_model_args = jit_model_args
         self._states = None
         self._states_flat = None
+        self._cache = {}
 
-    def _single_chain_mcmc(self, init, collect_fields=('z',), collect_warmup=False, args=(), kwargs={}):
-        rng_key, init_params = init
-        init_state, sample_fn, constrain_fn = self.sampler.init(rng_key, self.num_warmup, init_params,
-                                                                model_args=args, model_kwargs=kwargs)
-        if self.constrain_fn is None:
-            constrain_fn = identity if constrain_fn is None else constrain_fn
+    def _get_cached_fn(self):
+        if self._jit_model_args:
+            args, kwargs = (None,), (None,)
         else:
-            constrain_fn = self.constrain_fn
+            # XXX: Is there a better hash key that we can use?
+            def _hashable(x):
+                # When the arguments are JITed, ShapedArray is hashable.
+                if isinstance(x, Tracer):
+                    return x
+                elif isinstance(x, DeviceArray):
+                    return x.copy().tobytes()
+                elif isinstance(x, np.ndarray):
+                    return x.tobytes()
+                return x
+            args = tree_map(lambda x: _hashable(x), self._args)
+            kwargs = tree_map(lambda x: _hashable(x), tuple(sorted(self._kwargs.items())))
+        key = args + kwargs
+        try:
+            fn = self._cache.get(key, None)
+        except TypeError:
```

Comment:
Could you add a comment for this logic? I don't understand why `TypeError` happens here.

Reply:
Yes, that might happen when args/kwargs contains an unhashable type. In that case, we can't do any caching, but this shouldn't fail. I'll add a comment.

Reply:
Oh, I get it now. Thanks for explaining!